rosbag2_test_common/include/rosbag2_test_common/publication_manager.hpp

```diff
@@ -71,31 +72,90 @@ class PublicationManager
             std::this_thread::sleep_for(interval);
           }
         }
-      });
+      };
+
+    publishers_.push_back(std::make_pair(publisher, publisher_fcn));
   }
 
   void run_publishers(bool verbose = false) const
   {
-    for (const auto & pub_fcn : publishers_) {
-      pub_fcn(verbose);
+    for (const auto & pub : publishers_) {
+      std::get<PublicationF>(pub)(verbose);
     }
   }
 
   void run_publishers_async(bool verbose = false) const
   {
     std::vector<std::future<void>> futures;
-    for (const auto & pub_fcn : publishers_) {
-      futures.push_back(std::async(std::launch::async, pub_fcn, verbose));
+    for (const auto & pub : publishers_) {
+      futures.push_back(std::async(std::launch::async, std::get<PublicationF>(pub), verbose));
     }
     for (auto & publisher_future : futures) {
       publisher_future.get();
     }
   }
 
+  /// \brief Wait until publisher with specified topic will be connected to the subscribers or
+  ///    timeout occur.
+  /// \tparam Timeout Data type for timeout duration from std::chrono:: namespace
+  /// \param topic_name topic name to find corresponding publisher
+  /// \param timeout Maximum time duration during which discovery should happen.
+  /// \param n_subscribers_to_match Number of subscribers publisher should have for match.
+  /// \return true if not find publisher by topic name or publisher has specified number of
+  ///    subscribers, otherwise false.
+  template<typename Timeout>
+  bool wait_for_matched(
+    const char * topic_name,
+    Timeout timeout, size_t n_subscribers_to_match = 1)
+  {
+    rclcpp::PublisherBase::SharedPtr publisher = nullptr;
+    for (const auto pub : publishers_) {
+      if (!std::strcmp(
+          std::get<rclcpp::PublisherBase::SharedPtr>(pub)->get_topic_name(),
+          topic_name))
+      {
+        publisher = std::get<rclcpp::PublisherBase::SharedPtr>(pub);
+        break;
+      }
+    }
+
+    if (publisher == nullptr) {
+      return false;
+    }
+
+    if (publisher->get_subscription_count() +
+      publisher->get_intra_process_subscription_count() >= n_subscribers_to_match)
+    {
+      return true;
+    }
+
+    auto sleep_time = std::chrono::milliseconds(100);
```

Comment:
I would recommend to reduce default sleep timeout. I would chose some number from 1 to 10 mSec.

Reply:
Okay. I change to use 10ms as default.